crates/cubecl-wgpu/src/backend/vulkan.rs

```diff
@@ -0,0 +1,275 @@
+    unsafe {
+        adapter.as_hal::<hal::api::Vulkan, _, _>(|hal_adapter| {
+            if let Some(adapter) = hal_adapter {
+                register_features(adapter, props, features, comp_options);
+            }
+        })
+    }
+}
+
+/// Request device with required features, plus CMMA if available.
+fn request_device(
+    wgpu_adapter: &wgpu::Adapter,
+    adapter: &vulkan::Adapter,
+    mut features: Features,
+    limits: Limits,
+) -> (wgpu::Device, wgpu::Queue) {
+    let full_feat = features;
+    // This registers only f16 but not u8/i8, so remove so we can manually add them
+    features.remove(Features::SHADER_F16);
+    // Skip float features since we already register a more general version manually
+    features.remove(Features::SHADER_FLOAT32_ATOMIC);
+
+    let ash = adapter.shared_instance();
+    let mut extended_feat = ExtendedFeatures::from_adapter(ash.raw_instance(), adapter, features);
+    let extensions = adapter.required_device_extensions(features);
+    let mut phys_features = adapter.physical_device_features(&extensions, features);
+
+    let supported_feat = unsafe {
+        ash.raw_instance()
+            .get_physical_device_features(adapter.raw_physical_device())
+    };
+
+    let family_index = 0; //TODO
+    let family_info = DeviceQueueCreateInfo::default()
+        .queue_family_index(family_index)
+        .queue_priorities(&[1.0]);
+    let family_infos = [family_info];
+
+    let str_pointers = extended_feat
+        .extensions
+        .iter()
+        .map(|&s| {
+            // Safe because `device_extensions` entries have static lifetime.
+            s.as_ptr()
+        })
+        .collect::<Vec<_>>();
+
+    let pre_info = DeviceCreateInfo::default()
+        .queue_create_infos(&family_infos)
+        .enabled_extension_names(&str_pointers);
+    let mut info = phys_features.add_to_device_create(pre_info);
+    info = info.enabled_features(&supported_feat);
+    info = extended_feat.add_to_device_create(info);
+
+    let vk_device = unsafe {
+        ash.raw_instance()
+            .create_device(adapter.raw_physical_device(), &info, None)
+            .expect("Failed to create Vulkan device")
+    };
+
+    let device = unsafe {
+        adapter
+            .device_from_raw(
+                vk_device,
+                None,
+                &extensions,
+                full_feat,
+                &wgpu::MemoryHints::MemoryUsage,
+                family_info.queue_family_index,
+                0,
+            )
+            .expect("Failed to create HAL device")
+    };
+
+    let descriptor = DeviceDescriptor {
+        label: None,
+        required_features: full_feat,
+        required_limits: limits,
+        // The default is MemoryHints::Performance, which tries to do some bigger
+        // block allocations. However, we already batch allocations, so we
+        // can use MemoryHints::MemoryUsage to lower memory usage.
+        memory_hints: wgpu::MemoryHints::MemoryUsage,
```

Comment:
The `memory_hints` value should probably be extracted into a variable, since it is also used in the `adapter.create_from_raw` function call. I guess it is important for both to be the same